### PR TITLE
manually unpack txid in txin

### PR DESCRIPTION
**Pull request by redxaxder**

### CODE REVIEW

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs [hunk 1]
```diff
@@ -445,31 +457,69 @@ type TransTxId (c :: Type -> Constraint) era =
...
-data TxIn crypto = TxInCompact !(TxId crypto) {-# UNPACK #-} !Word64
-  deriving (Generic)
+data TxIn crypto where
+  TxInCompact32 ::
+    HS.SizeHash (CC.HASH crypto) ~ 32 =>
+    {-# UNPACK #-} !Word64 -> -- Hash part 1/4
+    {-# UNPACK #-} !Word64 -> -- Hash part 2/4
+    {-# UNPACK #-} !Word64 -> -- Hash part 3/4
+    {-# UNPACK #-} !Word64 -> -- Hash part 4/4
+    {-# UNPACK #-} !Word64 -> -- Index
```
Comment (dcoutts) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs:467]: Yes, this is clearly 6 words, so it's a win.

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs [hunk 2]
```diff
@@ -431,7 +440,10 @@ deriving newtype instance CC.Crypto crypto => FromCBOR (TxId crypto)
 deriving newtype instance CC.Crypto crypto => NFData (TxId crypto)
 
 instance HeapWords (TxIn crypto) where
-  heapWords (TxInCompact txid ix) = 3 + HW.heapWordsUnpacked txid + HW.heapWordsUnpacked ix
+  heapWords (TxInCompact32 a _ _ _ ix) =
+    6 + (4 * HW.heapWordsUnpacked a) + HW.heapWordsUnpacked ix
```
Comment (dcoutts) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs:444]: Are we sure that `heapWordsUnpacked (a :: Word64)` is zero? The answer here must be 6, so it must amount to 6 + 4 * 0 + 0. Can someone verify that the answer here does indeed come out to be simply 6?

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs [hunk 3]
```diff
@@ -445,31 +457,69 @@ type TransTxId (c :: Type -> Constraint) era =
...
-  TxInCompact addr (fromIntegral -> index)
+pattern TxIn txid index <-
+  (viewTxIn -> (txid, index))
   where
-    TxIn addr index =
-      TxInCompact addr (fromIntegral index)
+    TxIn txid@(TxId sh) index =
+      case HS.viewHash32 (extractHash sh) of
+        HS.ViewHashNot32 -> TxInCompactOther txid (fromIntegral index)
+        HS.ViewHash32 a b c d -> TxInCompact32 a b c d (fromIntegral index)
```
Comment (dcoutts) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs:482]: Noice.

**Review by dcoutts — APPROVED** [review]

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs [hunk 4]
```diff
@@ -445,31 +457,69 @@ type TransTxId (c :: Type -> Constraint) era =
...
+viewTxIn :: TxIn crypto -> (TxId crypto, Natural)
+viewTxIn (TxInCompactOther txid i) = (txid, fromIntegral i)
+viewTxIn (TxInCompact32 a b c d i) = (txid, fromIntegral i)
+  where
+    txid = TxId (unsafeMakeSafeHash $ HS.unsafeMkHash32 a b c d)
+
+instance Show (TxIn crypto) where
+  showsPrec d (viewTxIn -> (txid, ix)) =
+    showParen (d > app_prec) $
+      showString "TxId "
```
Comment (JaredCorduan) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs]:
```suggestion
      showString "TxIn "
```

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs (hunk 2 above)
Comment (redxaxder) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/TxBody.hs:444]: We have
```
instance HeapWords Word64 where
  heapWords _ = 2
```
and
```
heapWordsUnpacked :: HeapWords a => a -> Int
heapWordsUnpacked x = heapWords x - 2
```

**Pull request merged**